pkg/webhook/bootstrap.go

```diff
@@ -360,3 +352,100 @@ func (s *Server) service() runtime.Object {
 	}
 	return svc
 }
+
+func (s *Server) deployment() runtime.Object {
+	namespace := "default"
+	if s.Secret != nil {
+		namespace = s.Secret.Namespace
+	}
+
+	labels := map[string]string{
+		"app": "webhook-server",
+	}
+	if s.Service != nil {
+		for k, v := range s.Service.Selectors {
+			labels[k] = v
+		}
+	}
+
+	image := "webhook-server:latest"
+	if len(os.Getenv("IMG")) != 0 {
+		image = os.Getenv("IMG")
+	}
+
+	readOnly := false
+	if s.Service != nil {
+		readOnly = true
+	}
+
+	var volumeSource corev1.VolumeSource
+	if s.Secret != nil {
+		var mode int32 = 420
+		volumeSource = corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				DefaultMode: &mode,
+				SecretName:  s.Secret.Name,
+			},
+		}
+	} else {
+		volumeSource = corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
+		}
+	}
+
+	return &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "apps/v1",
+			Kind:       "Deployment",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "webhook-server",
+			Namespace: namespace,
+			Labels:    labels,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            "webhook-server-container",
+							Image:           image,
+							ImagePullPolicy: corev1.PullAlways,
+							Command: []string{
+								"/bin/sh",
+								"-c",
+								"exec ./manager --disable-installer",
+							},
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "webhook-server",
+									Protocol:      corev1.ProtocolTCP,
+									ContainerPort: s.Port,
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "cert",
+									MountPath: s.CertDir,
+									ReadOnly:  readOnly,
+								},
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name:         "cert",
+							VolumeSource: volumeSource,
+						},
+					},
+				},
+			},
+		},
+	}
+}
```

Comment:
I was thinking about just printing out the webhook config only. Outputting other manifests means, now we have two different ways to generate the manifests for controller ?

Reply:
IMO using generated Deployment (can also be a StatefulSet with a headless Service) is less error-prone.
There are quite some places need to be set correctly in order to make the webhook server binary work correctly.
We can discuss in person tomorrow :)